### PR TITLE
feat(app): configure payload length of CAMs via app

### PR DESCRIPTION
**Pull request by kschrab**
## Type of this PR 

- [ ] Bug fix
- [x] Enhancement

## Description

In previous versions the minimal length of CAM payload was configured globally via the `application_config.json`. This is not always suitable. With this PR, the minimal length of CAM payload is configured during activation of the communication module, e.g.:

```
getOs().getAdHocModule().enable(
        new AdHocModuleConfiguration()
                .camMinimalPayloadLength(200)
                .addRadio().channel(AdHocChannel.CCH).power(50).create()
);
```

If not configured, a default value of `200` Bytes is used.

Further changes:

* The minimal payload length of DENM, IVIM, and SPATM must be passed via its contructors.
* The `GenericV2xMessage` was used in old versions of VSimRTI to wrap around messages. This bit of code was not required anymore since the first release of MOSAIC. Instead of throwing the message away, I decided to change its purpose to a generic implementation of `V2xMessage`, which now can take any kind of payload, or no explicit payload at all, so that other applications not really need to implement own V2xMessage sub-classes. E.g. for just sending a message with a specific length a specific name, one could just use `new GenericV2xMessage(routing, name, length)`. In addition, the FileOutput can extract the name/type of the message for logging purposes.

## Issue(s) related to this PR

 * Internal issue 340 
 
## Affected parts of the online documentation

Yes, https://www.eclipse.org/mosaic/docs/simulators/application_simulator/ must be adjusted.

## Definition of Done

- [x] You have read CONTRIBUTING.md carefully.
- [x] You have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] All commits are signed-off (`-s`) with your mail address of your Eclipse Account.
- [x] `origin/main` has been merged into your Fork.
- [x] New functionality is covered by unit tests or integration tests. Code coverage must not decrease.
- [x] There are no new SpotBugs warnings. 
- [x] Coding guidelines have been observed (see CONTRIBUTING.md).
- [x] All tests pass.

### CODE REVIEW

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AbstractCommunicationModule.java [hunk 1]
```diff
@@ -142,7 +144,10 @@ void logEnableConfigurationNull() {
      */
     protected Integer sendCam(MessageRouting routing) {
         final CamBuilder camBuilder = owner.assembleCamMessage(new CamBuilder());
-        final Cam cam = new Cam(routing, camBuilder.create(owner.getSimulationTime(), owner.getId()));
+        final Cam cam = new Cam(routing,
+                camBuilder.create(owner.getSimulationTime(), owner.getId()),
+                ObjectUtils.defaultIfNull(configuration.camPayloadMinimalLength, 200L)
```
Comment (schwepmo) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AbstractCommunicationModule.java:149]: We could consider adding a constant for the `200L`

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AbstractCommunicationModule.java [hunk 2]
```diff
@@ -37,10 +37,13 @@
  * (Can be either a {@link AdHocModule} or a {@link CellModule})
  */
 @SuppressWarnings("checkstyle:ClassTypeParameterName")
-public abstract class AbstractCommunicationModule<ConfigT extends CommunicationModuleConfiguration> implements CommunicationModule<ConfigT> {
+public abstract class AbstractCommunicationModule<ConfigT extends AbstractCommunicationModuleConfiguration>
+        implements CommunicationModule<ConfigT> {
 
     private final AtomicInteger sequenceNumberGenerator;
 
+    protected ConfigT configuration;
```
Comment (schwepmo) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AbstractCommunicationModule.java:45]: 👍 for further abstraction

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AdHocModule.java [hunk 3]
```diff
@@ -63,38 +57,37 @@ public void enable() {
...
-                    interfaceBuilder.secondChannel(radio.getChannel1());
-                }
-                builder.addInterface(interfaceBuilder.create());
-            }
+            return;
+        }
 
-            owner.sendInteractionToRti(new AdHocCommunicationConfiguration(owner.getSimulationTime(), builder.create()));
+        AdHocConfiguration.Builder builder = new AdHocConfiguration.Builder(owner.getId());
+        for (int i = 0; i < min(2, configuration.getNrOfRadios()); i++) {
```
Comment (schwepmo) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AdHocModule.java:67]: Any reasons why we don't use a for each loop here? Would also allow us to get rid of the somewhat unnecessary method `getNrOfRadios()`

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/app/etsi/AbstractCamSendingApp.java [hunk 4]
```diff
@@ -139,7 +142,7 @@ private boolean dataChanged() {
 
         Data newData = generateEtsiData();
         if (newData == null) {
-            getLog().infoSimTime(this, "Could not check delta. Data are not available yet.");
+            getLog().debugSimTime(this, "Could not check delta. Data is not available yet.");
```
Comment (schwepmo) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/app/etsi/AbstractCamSendingApp.java:145]: Actually the word "data" is the plural of "datum" so using "are" was fine. But either way probably works

**Review by schwepmo — COMMENTED** [review]
Looks good overall. Left some minor remarks

**Comment by rprotzmann** [comment]
The realization is generally very well designed. I would merely rename every occurence of similar variables to `minimalPayloadLength` for unity with all messages

File: bundle/src/assembly/resources/scenarios/HelloWorld/application/EtsiApplication.json [hunk 5]
```diff
@@ -1,4 +1,5 @@
 {
+    "minPayloadLength": "200 Bytes",
```
Comment (rprotzmann) [inline bundle/src/assembly/resources/scenarios/HelloWorld/application/EtsiApplication.json:2]: `minimalPayloadLength`

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AbstractCommunicationModule.java [hunk 6]
```diff
@@ -104,34 +109,33 @@ public void sendV2xMessage(final V2xMessage msg) {
...
-        log.warn("Configuration provided to enable CommunicationModule is null");
+    @Override
+    public void enable(ConfigT configuration) {
+        if (configuration == null) {
+            log.warn("Configuration provided to enable CommunicationModule is null");
+            return;
+        }
+        if (this.configuration != null) {
+            log.info("This communication module has already been activated.");
+            if (configuration.camPayloadMinimalLength == null) {
```
Comment (rprotzmann) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AbstractCommunicationModule.java:128]: `camMinimalPayloadLength`

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AbstractCommunicationModuleConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ * Contact: mosaic@fokus.fraunhofer.de
+ */
+
+package org.eclipse.mosaic.fed.application.ambassador.simulation.communication;
+
+import org.eclipse.mosaic.fed.application.app.api.communication.CommunicationModuleConfiguration;
+
+public abstract class AbstractCommunicationModuleConfiguration implements CommunicationModuleConfiguration {
+
+    Long camPayloadMinimalLength = null;
```
Comment (rprotzmann) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AbstractCommunicationModuleConfiguration.java:22]: `camMinimalPayloadLength`

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AdHocModuleConfiguration.java [hunk 8]
```diff
@@ -82,6 +81,12 @@ public int getNrOfRadios() {
         return radios.size();
     }
 
+    @Override
+    public AdHocModuleConfiguration camPayloadLength(long minimalLength) {
```
Comment (rprotzmann) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AdHocModuleConfiguration.java:85]: `minimalPayloadLength`

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/app/etsi/AbstractCamSendingApp.java [hunk 9]
```diff
@@ -79,7 +80,9 @@ public void onStartup() {
      */
     protected void activateCommunicationModule() {
         getOperatingSystem().getAdHocModule().enable(
-                new AdHocModuleConfiguration().addRadio().channel(AdHocChannel.CCH).power(50).create()
+                new AdHocModuleConfiguration()
+                        .camPayloadLength(getConfiguration().minPayloadLength / DATA.BYTE)
```
Comment (rprotzmann) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/app/etsi/AbstractCamSendingApp.java:84]: `minimalPayloadLength`

File: lib/mosaic-objects/src/main/java/org/eclipse/mosaic/lib/objects/v2x/etsi/Cam.java [hunk 10]
```diff
@@ -53,22 +53,21 @@ public EncodedPayload getPayLoad() {
      * @param routing    Message to be routed.
      * @param camContent Content of a CAM Message.
      */
-    public Cam(final MessageRouting routing, final CamContent camContent) {
+    public Cam(final MessageRouting routing, final CamContent camContent, long minimalCamLength) {
```
Comment (rprotzmann) [inline lib/mosaic-objects/src/main/java/org/eclipse/mosaic/lib/objects/v2x/etsi/Cam.java:56]: `minimalPayloadLength` for same method signature as with IVIMs, SPATMs, DENMs

**Review by rprotzmann — APPROVED** [review]

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AbstractCommunicationModule.java (hunk 1 above)
Comment (kschrab) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/communication/AbstractCommunicationModule.java]: done

**Pull request merged**